core/data/src/main/kotlin/com/google/samples/apps/nowinandroid/core/data/util/TimeZoneMonitor.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.samples.apps.nowinandroid.core.data.util
+
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.os.Build.VERSION
+import android.os.Build.VERSION_CODES
+import androidx.tracing.trace
+import com.google.samples.apps.nowinandroid.core.network.Dispatcher
+import com.google.samples.apps.nowinandroid.core.network.NiaDispatchers.IO
+import com.google.samples.apps.nowinandroid.core.network.di.ApplicationScope
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharedFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.conflate
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.shareIn
+import kotlinx.datetime.TimeZone
+import java.time.ZoneId
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Utility for reporting current timezone the device has set.
+ * It always emits at least once with default setting and then for each TZ change.
+ */
+interface TimeZoneMonitor {
+    val currentTimeZone: Flow<TimeZone>
+}
+
+@Singleton
+internal class TimeZoneBroadcastMonitor @Inject constructor(
+    @ApplicationContext private val context: Context,
+    @ApplicationScope appScope: CoroutineScope,
+    @Dispatcher(IO) private val ioDispatcher: CoroutineDispatcher,
+) : TimeZoneMonitor {
+
+    override val currentTimeZone: SharedFlow<TimeZone> =
+        callbackFlow {
+            // Send the default time zone first.
```

Comment:
Nit: this is probably extremely unlikely, but since we are registering the receiver asynchronously, and reporting the initial time zone first, I think there's technically a race condition if the timezone happens to change after the initial `trySend`, but before the receiver gets registered.

I think a fix would be to call this `trySend` after we register the receiver, but that might result in a longer length of time before the `callbackFlow` emits anything. It should be safe if we call `trySend` twice, once here and once after registering the receiver? And if we wanted to avoid duplicate emissions, we could add a `distinctUntilChanged`

Reply:
I added it there, but I think we're overcomplicating it a bit here 😀 .